src/java/org/apache/cassandra/db/ColumnFamilyStore.java

```diff
@@ -629,9 +659,30 @@ public void invalidate(boolean expectMBean)
         strategyContainer.shutdown();
         SystemKeyspace.removeTruncationRecord(metadata.id);
 
-        data.dropSSTables();
-        LifecycleTransaction.waitForDeletions();
-        indexManager.dropAllIndexes();
+        storageHandler.withReloadingDisabled(() -> {
+            if (dropData)
+            {
+                data.dropSSTables();
+                LifecycleTransaction.waitForDeletions();
+
+                indexManager.dropAllIndexes();
+            }
+            else
+            {
+                // In CNDB, because of multi-tenancy, we might just unload a CFS without deleting the data as
+                // a tenant can be moved to a different set of nodes, which will then need to read data from remote storage
+                data.unloadSSTables();
+
+                indexManager.unloadAllIndexes();
+            }
+
+            storageHandler.unload();
+
+            if (dropData)
+            {
+                LifecycleTransaction.waitForDeletions(); // just in case an index had a reference on the sstable
```

Comment:
Is there a reason not to just move the previous "wait for deletion" after "drop indexes"?  Also this was not a problem for the previous code, why is it a possible issue now?

Reply:
> Is there a reason not to just move the previous "wait for deletion" after "drop indexes"? 

good idea, the previous one can be removed.

>  Also this was not a problem for the previous code, why is it a possible issue now?

SAI is keeping the sstable references, so the sstable tidier runs after `dropAllIndexes()`, see https://datastax.jira.com/browse/DB-3575